ddht/v5_1/network.py

```diff
@@ -403,21 +404,67 @@ async def find_nodes(
         # Validate that all responses are indeed at one of the
         # specified distances.
         for response in responses:
-            for enr in response.message.enrs:
-                if enr.node_id == node_id:
-                    if 0 not in distances:
-                        raise ValidationError(
-                            f"Invalid response: distance=0  expected={distances}"
-                        )
-                else:
-                    distance = compute_log_distance(enr.node_id, node_id)
-                    if distance not in distances:
-                        raise ValidationError(
-                            f"Invalid response: distance={distance}  expected={distances}"
-                        )
+            validate_found_nodes_distances(response, node_id, distances)
 
         return tuple(enr for response in responses for enr in response.message.enrs)
 
+    @asynccontextmanager
+    async def stream_find_nodes(
+        self,
+        node_id: NodeID,
+        endpoint: Endpoint,
+        distances: Collection[int],
+        *,
+        request_id: Optional[bytes] = None,
+    ) -> AsyncIterator[trio.abc.ReceiveChannel[InboundMessage[FoundNodesMessage]]]:
+        if not distances:
+            raise TypeError("Must provide at least one distance")
+
+        if endpoint is None:
+            endpoint = await self.endpoint_for_node_id(node_id)
+
+        send_channel, receive_channel = trio.open_memory_channel[
+            InboundMessage[FoundNodesMessage]
+        ](256)
+
+        async with trio.open_nursery() as nursery:
+            nursery.start_soon(
+                self._stream_find_nodes_response,
+                node_id,
+                endpoint,
+                distances,
+                request_id,
+                send_channel,
+            )
+
+            try:
+                async with receive_channel:
+                    try:
+                        yield receive_channel
+                    except trio.EndOfChannel as err:
+                        raise trio.TooSlowError from err
+            finally:
+                nursery.cancel_scope.cancel()
+
+    async def _stream_find_nodes_response(
+        self,
+        node_id: NodeID,
+        endpoint: Endpoint,
+        distances: Collection[int],
+        request_id: Optional[bytes],
+        send_channel: trio.abc.SendChannel[InboundMessage[FoundNodesMessage]],
+    ) -> None:
+        async with self.client.stream_find_nodes(
+            node_id, endpoint, distances=distances, request_id=request_id
+        ) as resp_aiter:
+            async with send_channel:
+                async for response in resp_aiter:
+                    # Validate that all responses are indeed at one of the
+                    # specified distances.
+                    validate_found_nodes_distances(response, node_id, distances)
+
+                    await send_channel.send(response)
```

Comment:
Here you should be sending the individual ENR records over the channel as the `NetworkAPI.stream_found_nodes` should stream the individual ENR records.